sycl/doc/design/IfDevice.md

```diff
@@ -0,0 +1,179 @@
+trivial implementation that works in the current multi-pass compiler.  The
+other is the design that we will ultimately use in the 1-pass compiler.
+
+
+## Multi-pass compiler implementation
+
+This implementation requires changes only to the device headers.  The
+implementation is very trivial, leveraging the existing `__SYCL_DEVICE_ONLY__`
+macro which is defined differently in the host compiler pass vs. the device
+compiler passes.
+
+```
+namespace sycl::ext::oneapi::experimental {
+namespace detail {
+
+// Helper object used to implement "otherwise".  The "MakeCall" template
+// parameter tells whether the previous call to "if_device" or "if_host" called
+// its "fn".  When "MakeCall" is true, the previous call to "fn" did not
+// happen, so the "otherwise" should call "fn".
+template<bool MakeCall>
+class if_device_or_host_helper {
+ public:
+  template<typename T>
+  void otherwise(T &&fn) {
+    if constexpr (MakeCall) {
+      fn();
+    }
+  }
+};
+
+} // namespace detail
+
+template<typename T>
+static auto if_device(T &&fn) {
+#ifdef __SYCL_DEVICE_ONLY__
+  fn();
+  return detail::if_device_or_host_helper<false>{};
+#else
+  return detail::if_device_or_host_helper<true>{};
+#endif
+}
+
+template<typename T>
+static auto if_host(T &&fn) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return detail::if_device_or_host_helper<true>{};
+#else
+  fn();
+  return detail::if_device_or_host_helper<false>{};
+#endif
+}
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+
+## Single-pass compiler implementation
+
+This implementation requires changes to the device headers, some changes to
+the error diagnostics in the front-end (CFE), and a new IR pass.
+
+### Device headers
+
+The device headers translate the API into calls to two functions that are
+decorated with attributes named "sycl-call-if-on-device" and
+"sycl-call-if-on-host".
+
+```
+namespace sycl::ext::oneapi::experimental {
+namespace detail {
+
+// Call the callable object "fn" only when this code runs on a device.
+//
+// IR passes recognize this function from the "sycl-call-if-on-device"
+// attribute.
+template<typename T>
+[[clang::noinline]]
+[[__sycl_detail__::add_ir_attributes_function("sycl-call-if-on-device", true)]]
+void call_if_on_device(T &&fn) {
+  fn();
+}
+
+// Call the callable object "fn" only when this code runs on the host.
+//
+// IR passes recognize this function from the "sycl-call-if-on-host" attribute.
+template<typename T>
+[[clang::noinline]]
+[[__sycl_detail__::add_ir_attributes_function("sycl-call-if-on-host", true)]]
+void call_if_on_host(T &&fn) {
+  fn();
+}
+
+class call_if_on_device_helper {
+ public:
+  template<typename T>
+  void otherwise(T &&fn) {
+    call_if_on_device(fn);
+  }
+};
+
+class call_if_on_host_helper {
+ public:
+  template<typename T>
+  void otherwise(T &&fn) {
+    call_if_on_host(fn);
+  }
+};
+
+} // namespace detail
+
+template<typename T>
+static auto if_device(T &&fn) {
+  detail::call_if_on_device(fn);
```

Comment:
@rolandschulz 

I think I do *not* need to use `std::forward` here, correct?

Reply:
Without the `std::forward`, if a r-value-ref gets passed to `if_device` then `call_if_on_device` gets passed a l-value-ref. If the callable only works with an r-value-ref this break. Always use `forward` for any universal reference (if you don't use it anymore afterwards).

Reply:
Thanks.  Added in f3e15a7a3b45370db79c529e3b876329810c555e.